### PR TITLE
AggregateRoot's repository instrumentation

### PR DESCRIPTION
**Pull request by mpraglowski**
Instruments `load` & `store` methods. Tracks aggregate class, stream and
for storing also current aggregate version & number of events stored.

### CODE REVIEW

File: aggregate_root/lib/aggregate_root/instrumented_repository.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+                                 stream_name: stream_name) do
+        repository.load(aggregate, stream_name)
+      end
+    end
+
+    def store(aggregate, stream_name)
+      instrumentation.instrument("store.repository.aggregate_root",
+                                 aggregate_class: aggregate.class,
+                                 aggregate_version: aggregate.version,
+                                 stored_events: aggregate.unpublished_events.size,
```
Comment (pawelpacana) [inline aggregate_root/lib/aggregate_root/instrumented_repository.rb]: Perhaps we should pass just `aggregate.unpublished_events` here. That would be in symmetry with `append_to_stream.repository.rails_event_store`.

File: aggregate_root/lib/aggregate_root/instrumented_repository.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+                                 stream_name: stream_name) do
```
Comment (pawelpacana) [inline aggregate_root/lib/aggregate_root/instrumented_repository.rb]: s/stream_name/stream, again following `append_to_stream.repository.rails_event_store`

File: aggregate_root/lib/aggregate_root/instrumented_repository.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+                                 aggregate_class: aggregate.class,
+                                 stream_name: stream_name) do
+        repository.load(aggregate, stream_name)
+      end
+    end
+
+    def store(aggregate, stream_name)
+      instrumentation.instrument("store.repository.aggregate_root",
+                                 aggregate_class: aggregate.class,
+                                 aggregate_version: aggregate.version,
```
Comment (pawelpacana) [inline aggregate_root/lib/aggregate_root/instrumented_repository.rb]: s/aggregate_version/version/

File: aggregate_root/lib/aggregate_root/instrumented_repository.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      instrumentation.instrument("load.repository.aggregate_root",
+                                 aggregate_class: aggregate.class,
+                                 stream_name: stream_name) do
+        repository.load(aggregate, stream_name)
+      end
+    end
+
+    def store(aggregate, stream_name)
+      instrumentation.instrument("store.repository.aggregate_root",
+                                 aggregate_class: aggregate.class,
```
Comment (pawelpacana) [inline aggregate_root/lib/aggregate_root/instrumented_repository.rb]: I think passing the instance, labeled as aggregate would be fine (thus s/aggregate_class/aggregate/)

Comment (pawelpacana) [reply aggregate_root/lib/aggregate_root/instrumented_repository.rb]: Instrumentation is useful in logging context, being able to track particular instance is helping.

File: aggregate_root/spec/instrumented_repostory_spec.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
+require 'spec_helper'
+require 'active_support/notifications'
+
+module AggregateRoot
+  RSpec.describe InstrumentedRepository do
+
+    describe "#load" do
+      specify "wraps around original implementation" do
+        some_repository = spy
```
Comment (pawelpacana) [inline aggregate_root/spec/instrumented_repostory_spec.rb]: Is there a way for spy to check if it still follows an API that is mocked? I'm always wary of such tests. I briefly remember https://github.com/psyho/bogus implementing this check.

Comment (pawelpacana) [reply aggregate_root/spec/instrumented_repostory_spec.rb]: https://relishapp.com/rspec/rspec-mocks/v/3-8/docs/verifying-doubles/using-an-instance-double

File: aggregate_root/spec/instrumented_repostory_spec.rb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        subscribe_to("store.repository.aggregate_root") do |notification_calls|
+          aggregate = Order.new
+          aggregate.create
+          aggregate.expire
+
+          instrumented_repository.store(aggregate, 'SomeStream')
+
+          expect(notification_calls).to eq([{
+            aggregate_class: Order,
+            aggregate_version: -1,
```
Comment (pawelpacana) [inline aggregate_root/spec/instrumented_repostory_spec.rb]: Interesting question, are we more interested in version before or after storing events? Also what happens in context of instrumentation if we cannot store events?

Comment (mpraglowski) [reply aggregate_root/spec/instrumented_repostory_spec.rb]: Usually repository gets a new instance of aggregate, so on load method call version does not matter. The version in store instrumentation is an aggregate version **before** store.

**Pull request merged**